hedera-node/src/main/java/com/hedera/services/ledger/PureTransferSemanticChecks.java

```diff
@@ -0,0 +1,194 @@
+ */
+public class PureTransferSemanticChecks {
+	public ResponseCodeEnum fullPureValidation(
+			TransferList hbarAdjustsWrapper,
+			List<TokenTransferList> tokenAdjustsList,
+			ImpliedTransfersMeta.ValidationProps validationProps
+	) {
+		final var maxHbarAdjusts = validationProps.getMaxHbarAdjusts();
+		final var maxTokenAdjusts = validationProps.getMaxTokenAdjusts();
+		final var maxOwnershipChanges = validationProps.getMaxOwnershipChanges();
+
+		final var hbarAdjusts = hbarAdjustsWrapper.getAccountAmountsList();
+
+		if (hasRepeatedAccount(hbarAdjusts)) {
+			return ACCOUNT_REPEATED_IN_ACCOUNT_AMOUNTS;
+		}
+		if (!isNetZeroAdjustment(hbarAdjusts)) {
+			return INVALID_ACCOUNT_AMOUNTS;
+		}
+		if (!isAcceptableSize(hbarAdjusts, maxHbarAdjusts)) {
+			return TRANSFER_LIST_SIZE_LIMIT_EXCEEDED;
+		}
+
+		final var tokenValidity = validateTokenTransferSizes(tokenAdjustsList, maxTokenAdjusts, maxOwnershipChanges);
+		if (tokenValidity != OK) {
+			return tokenValidity;
+		}
+		return validateTokenTransferSemantics(tokenAdjustsList);
+	}
+
+	ResponseCodeEnum validateTokenTransferSizes(
+			List<TokenTransferList> tokenTransfersList,
+			int maxListLen,
+			int maxOwnershipChanges
+	) {
+		final int numScopedTransfers = tokenTransfersList.size();
+		if (numScopedTransfers == 0) {
+			return OK;
+		}
+
+		if (numScopedTransfers > maxListLen) {
+			return TOKEN_TRANSFER_LIST_SIZE_LIMIT_EXCEEDED;
+		}
+
+		var count = 0;
+		var numOwnershipChanges = 0;
+		for (var scopedTransfers : tokenTransfersList) {
+			final var ownershipChangesHere = scopedTransfers.getNftTransfersCount();
+			if (ownershipChangesHere > 0) {
+				numOwnershipChanges += ownershipChangesHere;
+				if (numOwnershipChanges > maxOwnershipChanges) {
+					return BATCH_SIZE_LIMIT_EXCEEDED;
+				}
+			} else {
+				int transferCounts = scopedTransfers.getTransfersCount();
+				if (transferCounts == 0) {
+					return EMPTY_TOKEN_TRANSFER_ACCOUNT_AMOUNTS;
+				}
+				count += transferCounts;
+				if (count > maxListLen) {
+					return TOKEN_TRANSFER_LIST_SIZE_LIMIT_EXCEEDED;
+				}
+			}
+		}
+
+		return OK;
+	}
+
+	ResponseCodeEnum validateTokenTransferSemantics(List<TokenTransferList> tokenTransfersList) {
+		if (tokenTransfersList.isEmpty()) {
+			return OK;
+		}
+		ResponseCodeEnum validity;
+		final Set<TokenID> uniqueTokens = new HashSet<>();
+		for (var tokenTransfers : tokenTransfersList) {
+			validity = validateScopedTransferSemantics(uniqueTokens, tokenTransfers);
+			if (validity != OK) {
+				return validity;
+			}
+		}
+		if (uniqueTokens.size() < tokenTransfersList.size()) {
+			return TOKEN_ID_REPEATED_IN_TOKEN_LIST;
+		}
+		return OK;
+	}
+
+	private ResponseCodeEnum validateScopedTransferSemantics(
+			Set<TokenID> uniqueTokens,
+			TokenTransferList tokenTransfers
+	) {
+		if (!tokenTransfers.hasToken()) {
+			return INVALID_TOKEN_ID;
+		}
+		uniqueTokens.add(tokenTransfers.getToken());
+		final var adjusts = tokenTransfers.getTransfersList();
```

Comment:
I think that a meaningful check to do for nft transfers would be checking that there is `sender`, `receiver` and `serialNumber` populated

Reply:
True, that would be consistent with the fungible prechecks...will add an `INCOMPLETE_NFT_TRANSFER` or similar response code in future PR. 👍